cocos/network/CCWebSocket-android.cpp

```diff
@@ -0,0 +1,384 @@
+/****************************************************************************
+ Copyright (c) 2019 Xiamen Yaji Software Co., Ltd.
+ http://www.cocos.com
+
+ Permission is hereby granted, free of charge, to any person obtaining a copy
+ of this software and associated documentation files (the "Software"), to deal
+ in the Software without restriction, including without limitation the rights
+ to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ copies of the Software, and to permit persons to whom the Software is
+ furnished to do so, subject to the following conditions:
+
+ The above copyright notice and this permission notice shall be included in
+ all copies or substantial portions of the Software.
+
+ THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ THE SOFTWARE.
+ ****************************************************************************/
+
+
+#include "network/CCWebSocket-android.h"
+#include "network/WebSocket.h"
+#include "platform/CCFileUtils.h"
+#include "platform/android/jni/JniHelper.h"
+
+
+#define JCLS_WEBSOCKET "org/cocos2dx/lib/Cocos2dxWebSocket"
+#define J_BINARY_CLS_WEBSOCKET "org.cocos2dx.lib.Cocos2dxWebSocket"
+#define JARG_STR        "Ljava/lang/String;"
+#define JARG_DOWNLOADER "L" JCLS_WEBSOCKET ";"
+
+#include <string>
+#include <mutex>
+#include <unordered_map>
+#include <condition_variable>
+#include <chrono>
+
+using namespace std;
+
+
+namespace {
+
+    struct ConditionVariable {
+        std::mutex mutex;
+        std::condition_variable cond;
+    };
```

Comment:
would it be better to implement `lock()` `notify()` in `ConditionVariable`?

Reply:
Yes